src/facade/cmd_arg_parser.cc

```diff
@@ -0,0 +1,67 @@
+// Copyright 2023, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "facade/cmd_arg_parser.h"
+
+#include <absl/strings/ascii.h>
+#include <absl/strings/match.h>
+#include <absl/strings/numbers.h>
+
+#include "base/logging.h"
+#include "facade/error.h"
+
+namespace facade {
+
+CmdArgParser::CheckProxy::operator bool() const {
+  if (idx_ >= parser_->args_.size())
+    return false;
+
+  std::string_view arg = parser_->SafeSV(idx_);
+  if ((!ignore_case_ && arg != tag_) || (ignore_case_ && !absl::EqualsIgnoreCase(arg, tag_)))
+    return false;
+
+  if (idx_ + expect_tail_ >= parser_->args_.size()) {
+    parser_->Report(SHORT_OPT_TAIL, idx_);
+    return false;
+  }
+
+  parser_->cur_i_++;
+
+  if (size_t uidx = idx_ + expect_tail_ + 1; next_upper_ && uidx < parser_->args_.size())
+    parser_->ToUpper(uidx);
+
+  return true;
+}
+
+template <typename T> T CmdArgParser::NextProxy::Int() {
+  T out;
+  if (absl::SimpleAtoi(operator std::string_view(), &out))
+    return out;
+  parser_->Report(INVALID_INT, idx_);
+  return T{0};
+}
+
+template uint64_t CmdArgParser::NextProxy::Int<uint64_t>();
```

Comment:
Unfortunately this doesn't work the way you expect it. 

First you don't test this, and that's why you are not getting a compilation error. If you use this function in another source file that includes the header `cmd_arg_parser.cc` it will fail to compile. 

You have two solutions here:
1. Move this to the header, let each translation unit explicitly instantiate this template and the linker fold its definition.

2. If you really expect these two cases to be used ONLY then you must add the following declarations to the header:

```
extern template uint64_t CmdArgParser::NextProxy::Int<uint64_t>;
extern template int64_t CmdArgParser::NextProxy::Int<int64_t>;
```

This will allow other translation units to wait for link time to resolve the symbol (which they will happily find since you explicitly instantiate it in this file)

Reply:
I use it in both `search_family` and `cmd_arg_parser_test.cc` and it compiles 🧙🏻‍♂️ 

AFAIK `extern template` would be necessarily to prevent instantiation on a defined template (your 3rd point actually). This template here just transforms into a declaration.

Reply:
Nope. 

1. It works because you explicitly instantiate the template in the `.cc` file -- you are not declaring anything there, you explicitly instantiate it. You are unlucky, because `size_t` is a type alias for `uint64_t` in your system so you are not getting any error. Try that out with another type, like `int` or something else, it will break. Also, try removing the explicit instantiation from your source file it will break again, the linker will complain.

> AFAIK extern template would be necessarily to prevent instantiation on a defined template (your 3rd point actually).

Yes! 

Is it meant to only be used for `uint64_t` && `int64_t` ?

Reply:
Yes, its ment to be used only with those types